feat(jans-cedarling): create uniffi binding for cedarling with sample ios app using it

duttarnab opened the pull request:
closes #9432
### Prepare

- [x] Read [PR guidelines](https://github.com/JanssenProject/jans/blob/main/docs/CONTRIBUTING.md#prs)
- [x] Read [license information](https://github.com/JanssenProject/jans/blob/main/LICENSE)

-------------------

### Description

#### Target issue
  <!-- Link or describe the issue this PR is fixing -->
  
  <!-- If issue shouldn't be closed after merging this PR, then we recommend adding a task in original target issue and create a separate issue from this task which can be closed when this PR gets merged. Mention this new issue created from task as target issue below. For more on how to create task issues visit https://docs.github.com/en/issues/tracking-your-work-with-issues/about-task-lists -->
  
closes #issue-number-here

#### Implementation Details
  <!-- If the fix is an involved one then communicate high level analysis and implementation approach -->

-------------------
### Test and Document the changes
- [x] Static code analysis has been run locally and issues have been fixed
- [x] Relevant unit and integration tests have been added/updated
- [x] Relevant documentation has been updated if any (i.e. user guides, installation and configuration guides, technical design docs etc)



Please check the below before submitting your PR. The PR will not be merged if there are no commits that start with `docs:` to indicate documentation changes or if the below checklist is not selected.
- [ ] **I confirm that there is no impact on the docs due to the code changes in this PR.**

rmarinn commented on jans-cedarling/bindings/cedarling_native_apps/README.md:
Could you add an Android section and indicate that it's a work in progress? This will make the README clearer, as it currently states that the project is for both iOS and Android but does not include both.

duttarnab replied:
done

rmarinn commented on jans-cedarling/bindings/cedarling_native_apps/iOSApp/iOSApp/context.json:
this looks like an empty file, is this needed?

duttarnab replied:
yes. We can test the app by adding the context data in this file.yes. We can test the app by adding the context data in this file.

rmarinn replied:
okay

nynymike reviewed (commented):
I see the Readme, but is there an impact on the docs? 

Docs Added

nynymike commented:
I'd like @NazarYavornytskyy to review this PR, as he's the most experienced iOS developer on our team.

olehbozhok commented:
Actually I don't really understand name `jcedarling_native_apps` because it is not native implementation it is `uniffi` implementation.

olehbozhok commented:
Also, I see a lot of `print` in code. And I can't believe it is good approach for logging.

olehbozhok commented on jans-cedarling/bindings/cedarling_native_apps/src/lib.rs:
```diff
@@ -0,0 +1,240 @@
+        request_id: &str,
+        tag: &str,
+    ) -> Result<Vec<String>, LogError> {
+        let mut result = Vec::new();
```
We can allocate result with defined capacity. With length from result `self.inner.get_logs_by_request_id_and_tag(request_id, tag)`

duttarnab replied:
done

olehbozhok commented on jans-cedarling/bindings/cedarling_native_apps/src/lib.rs:
```diff
@@ -0,0 +1,240 @@
+    /// Return log entries that match the given request_id.
+    #[uniffi::method]
+    pub fn get_logs_by_request_id(&self, request_id: &str) -> Result<Vec<String>, LogError> {
+        let mut result = Vec::new();
```
We can allocate result with defined capacity

duttarnab replied:
done

olehbozhok commented on jans-cedarling/bindings/cedarling_native_apps/src/lib.rs:
```diff
@@ -0,0 +1,240 @@
+    /// Tag can be `log_kind`, `log_level`.
+    #[uniffi::method]
+    pub fn get_logs_by_tag(&self, tag: &str) -> Result<Vec<String>, LogError> {
+        let mut result = Vec::new();
```
We can allocate result with defined capacity

duttarnab replied:
done

olehbozhok commented on jans-cedarling/bindings/cedarling_native_apps/src/lib.rs:
```diff
@@ -0,0 +1,240 @@
+    //Get all log ids
+    #[uniffi::method]
+    pub fn get_log_ids(&self) -> Vec<String> {
+        let mut result = Vec::new();
```
We can allocate result with defined capacity

duttarnab replied:
done

olehbozhok commented on jans-cedarling/bindings/cedarling_native_apps/src/lib.rs:
```diff
@@ -0,0 +1,240 @@
+    // Retrieves logs and serializes them as JSON strings
+    #[uniffi::method]
+    pub fn pop_logs(&self) -> Result<Vec<String>, LogError> {
+        let mut result = Vec::new();
```
We can allocate result with defined capacity

duttarnab replied:
done

olehbozhok commented on jans-cedarling/bindings/cedarling_native_apps/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+crate_type = ["cdylib", "staticlib"]
+name = "mobile"
+ 
+[dependencies]
```
To avoid error in compile `wasm` tests you can use 
```
# dependency for NOT wasm target
[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
```
instead of `[dependencies]`

like [here](https://github.com/JanssenProject/jans/blob/main/jans-cedarling/bindings/cedarling_python/Cargo.toml#L13C37-L13C45)


and don't forget to add [this](https://github.com/JanssenProject/jans/blob/main/jans-cedarling/bindings/cedarling_python/src/lib.rs#L7) in `lib.rs`

duttarnab replied:
done

olehbozhok commented on jans-cedarling/bindings/cedarling_native_apps/src/lib.rs:
```diff
@@ -0,0 +1,241 @@
+
+// Struct to hold authorization result, compatible with iOS serialization
+#[derive(Debug, serde::Serialize, uniffi::Record)]
+pub struct AuthorizeResult {
```
I don't see how user can get `request_id`
https://github.com/JanssenProject/jans/blob/main/jans-cedarling/cedarling/src/authz/authorize_result.rs#L35C9-L35C19

duttarnab replied:
done

duttarnab commented:
> Actually I don't really understand name `jcedarling_native_apps` because it is not native implementation it is `uniffi` implementation.

Renamed to `cedarling_uniffi`.

>  Also, I see a lot of print in code. And I can't believe it is good approach for logging.

Removed unnecessary print statements.

> It is supposed to be Swift application or only lib to import to real application?

The binding can be used in  Python, Swift or Kotlin code, but currently it has been tested only for swift.

olehbozhok commented on jans-cedarling/bindings/cedarling_uniffi/src/request_wrapper.rs:
```diff
@@ -0,0 +1,71 @@
+    ) -> Result<Self, RequestError> {
+        // Validate tokens: ensure all values are non-empty after trimming
+        let mut validated_tokens = HashMap::new();
+        for (key, value) in tokens.iter() {
```
Is it possible to use `tokens.into_iter()`, and avoid cloning of `key`?

duttarnab replied:
done

olehbozhok reviewed (approved):
Looks OK to me.

olehbozhok commented:
You can use next code snippet to avoid cargo error

bindings/cedarling_uniffi/src/bin/uniffi-bindgen.rs
```
#[cfg(not(target_arch = "wasm32"))]
fn main() {
    uniffi::uniffi_bindgen_main()
}

// just to avoid clippy error about missing `main` function
#[cfg(target_arch = "wasm32")]
fn main() {
}

```
